internal/pkg/v1/action/operator_update.go

```diff
@@ -0,0 +1,119 @@
+package action
+
+import (
+	"context"
+	"fmt"
+	"maps"
+	"slices"
+	"time"
+
+	"github.com/blang/semver/v4"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	olmv1 "github.com/operator-framework/operator-controller/api/v1"
+)
+
+type updateClient interface {
+	updater
+	getter
+}
+
+type OperatorUpdate struct {
+	client updateClient
+
+	Package string
+
+	Version                 string
+	Channels                []string
+	UpgradeConstraintPolicy string
+	Labels                  map[string]string
+	OverrideUnset           bool
+
+	CleanupTimeout time.Duration
+
+	Logf func(string, ...interface{})
+}
+
+func NewOperatorUpdate(client updateClient) *OperatorUpdate {
+	return &OperatorUpdate{
+		client: client,
+		Logf:   func(string, ...interface{}) {},
+	}
+}
+
+func (ou *OperatorUpdate) Run(ctx context.Context) (*olmv1.ClusterExtension, error) {
+	var ext olmv1.ClusterExtension
+
+	opKey := types.NamespacedName{Name: ou.Package}
+	if err := ou.client.Get(ctx, opKey, &ext); err != nil {
+		return nil, err
+	}
+
+	if ext.Spec.Source.SourceType != olmv1.SourceTypeCatalog {
+		return nil, fmt.Errorf("unrecognized source type: %q", ext.Spec.Source.SourceType)
+	}
+
+	ou.setDefaults(ext)
+	constraintPolicy := olmv1.UpgradeConstraintPolicy(ou.UpgradeConstraintPolicy)
+	if !ou.needsUpdate(ext, constraintPolicy) {
+		return nil, ErrNoChange
+	}
+
+	if ou.Version != "" {
+		if _, err := semver.ParseRange(ou.Version); err != nil {
+			return nil, fmt.Errorf("failed parsing version: %w", err)
+		}
+	}
+
+	ext.SetLabels(ou.Labels)
+	ext.Spec.Source.Catalog.Version = ou.Version
+	ext.Spec.Source.Catalog.Channels = ou.Channels
+	ext.Spec.Source.Catalog.UpgradeConstraintPolicy = constraintPolicy
+	if err := ou.client.Update(ctx, &ext); err != nil {
+		return nil, err
+	}
+
+	if err := waitUntilOperatorStatusCondition(ctx, ou.client, &ext, olmv1.TypeInstalled, metav1.ConditionTrue); err != nil {
+		return nil, fmt.Errorf("timed out waiting for operator: %w", err)
+	}
+
+	return &ext, nil
+}
+
+func (ou *OperatorUpdate) setDefaults(ext olmv1.ClusterExtension) {
+	catalogSrc := ext.Spec.Source.Catalog
+	if ou.OverrideUnset {
+		if ou.Version == "" {
+			ou.Version = catalogSrc.Version
+		}
+		if len(ou.Channels) == 0 {
+			ou.Channels = catalogSrc.Channels
+		}
+		if ou.UpgradeConstraintPolicy == "" {
+			ou.UpgradeConstraintPolicy = string(catalogSrc.UpgradeConstraintPolicy)
```

Comment:
What exactly means:

> fs.StringVar(&i.UpgradeConstraintPolicy, "upgrade-constraint-policy", "", "controls whether the upgrade path(s) defined in the catalog are enforced. One of CatalogProvided|SelfCertified), Default: CatalogProvided")

Is SelfCertified Marketplace something like that?
What does `whether the upgrade path(s) defined in the catalog are enforced` mean?
Could you please provide an example? Link for doc where we do that in OLMv1 or use case to better clarify what it means?

Reply:
> What does whether the upgrade path(s) defined in the catalog are enforced mean?
Could you please provide an example or use case to clarify what it means?
What is the difference?

this was taken verbatim from our CRD documentation. I'd rather not paste it here, because it's not brief, but hopefully this will be satisfactory for you to move forward? 

Please remember that we are planning on adjusting the docs once we have all commands (not that I think we should do that necessarily if it comes from our CRD docs)

Reply:
I found the explaination here; https://operator-framework.github.io/operator-controller/api-reference/operator-controller-api-reference/#catalogfilter

Cool,